src/csharp/Microsoft.Spark/Utils/PythonSerDe.cs

```diff
@@ -67,7 +68,23 @@ internal static object[] GetUnpickledObjects(Stream stream, int messageLength)
                     stackCapacity: 102); // Spark sends batches of 100 rows, and +2 is for markers.
                 s_rowConstructor.Reset();
                 Debug.Assert(unpickledItems != null);
-                return (unpickledItems as object[]);
+                // Cast unpickledItems if necessary.
+                foreach (object objArr in (object[])unpickledItems)
+                {
+                    if (objArr.GetType() == typeof(object[]))
```

Comment:
Can you add a detailed comment what's happening in this block? Also, would this incur perf hit since you are inspecting every row coming in?

Reply:
I updated the comments. I am adding the `foreach` loop to deal with `null` case. Basically, this block is to check if `unpickledItems` contains `ArrayList`, if so then we need to cast it to the appropriate array type using `UnpickleArray`. I added the `break` in the loop for the perf concern. Please advise and let me know if this makes sense. Thanks!

Reply:
What if there are 1000 objects in `unpickledItems` and 1000th one is `ArrayList`?

Reply:
Maybe the better question is https://github.com/dotnet/spark/pull/536/files#r475892642

Reply:
Please see the reply below. Thanks!